### DISCUSSION TITLE
[RFC-4] Subscriptions

### DISCUSSION THREAD

**Discussion by Amzd** [maticzav/swift-graphql#15] [General]
See prototype here: https://github.com/maticzav/swift-graphql/compare/main...Amzd:main

# Web sockets

I've done some research and to work with subscriptions you need to use web sockets. 

There is three big web socket libraries;

1. SocketRocket (archived)
2. Starscream
3. Apple's WebSocket support

SocketRocket has not been updated in 5 years so I think that's a no. 

Starscream gets updated regularly and apollo uses this too. It looks ok from initial checking.

Apple's WebSocket support is iOS 13 and up, since I no longer develop new features for iOS 12 this is not an issue for me so I went ahead and created a working prototype with this.

---

# Changes

This is changes I propose to make

1. Compile time access to operation type
2. Adding `SwiftGraphQL.observe` or `SwiftGraphQL.listen` (need feedback on naming) which allows for cancelling the websocket
3. A way to set availability of an operation (since in my example I use Apples web sockets which are iOS 13+)

## Compile time access to operation type

Right now we have the operation type as an enum. I propose changing that to an actual type so we can type check against it with generics.

Removing: 

**Document.swift**
```swift
public enum GraphQLOperationType: String, Codable, CaseIterable {
    case query = "query"
    case mutation = "mutation"
    case subscription = "subscription"
}
```

Adding:

**Operation.swift**
```swift
public protocol GraphQLOperation {
    static var operation: String { get }
}
public protocol GraphQLQuery: GraphQLOperation {}
public protocol GraphQLMutation: GraphQLOperation {}
public protocol GraphQLSubscription: GraphQLOperation {}

extension GraphQLQuery {
    public static var operation: String { "query" }
}
extension GraphQLMutation {
    public static var operation: String { "mutation" }
}
extension GraphQLSubscription {
    public static var operation: String { "subscription" }
}
```

In CodeGen we need to change the type:

**CodeGen/Operation.swift**
```swift
/* Operation*/
code.append("extension Operations.\(identifier): GraphQL\(identifier) {")
code.append("}")
code.append("")
```

What this enables is type checking for the type of operation so you cannot call SG.observe() with a Query.

```swift
public static func observe<Type, TypeLock>(
   // ...
) -> Token where TypeLock: GraphQLSubscription & Decodable {
//                         ^^^^^^^^^^^^^^^^^^^
```

## Adding `SwiftGraphQL.observe` or `SwiftGraphQL.listen` (need feedback on naming) which allows for cancelling the websocket

My SG.observe() builds a query similar to SG.send, but then sends it into a web socket which calls the onEvent every time an event happens.

Since you don't want this to happen indefinitely I have a Token object that when released from memory stops the web socket:

```swift
public static func observe<Type, TypeLock>(
   // ...
) -> Token where TypeLock: GraphQLSubscription & Decodable {
    // ...
    return Token {
        socket.cancel(with: .goingAway, reason: WebSocketCloseReason.tokenDeinit)
    }
}

public class Token {
    private var onDeinit: () -> Void
    internal init(onDeinit: @escaping () -> Void) {
        self.onDeinit = onDeinit
    }
    deinit {
        onDeinit()
    }
}
``` 

## A way to set availability of an operation (since in my example I use Apples web sockets which are iOS 13+)

I replaced the operation parameter in GraphQLCodegen.generateOperation with availibity that takes an optional String which I add before each extension if present.

```swift
func generateOperation(
    _ identifier: String,
    for type: GraphQL.ObjectType,
    availability: String?
) throws -> [String] {
    // ...
    /* Definition*/
    availability.map { code.append($0) }
    code.append("extension Operations {")
    // ...
```

**Comment by maticzav** [comment]
Thank you!! It's even better than I envisioned it.

- `Apple's WebSocket support` - perfect! I'd have suggested the same thing.
- `Compile time access to operation type` - let's check if we generate anything else in operations (besides what we do with objects). If not, let's reuse objects instead of operations (I can also help with this), and only generate the extensions you posted above in the #operations section. (If they provide the same level of type-safety).
- `SwiftGraphQL.listen` - let's go with `listen` seems more approachable.
- `generateOperation` - 💯 (let's move this to object if possible - related to compile-time access)


I suggest we move this conversation to a PR from the "demo". It seems like 80% of it is working from what I've seen. 

1. I'd remove additional `send` utility function for subscriptions - I think this is an unnecessary optimisation (let me know if you had any particular reason to include it).
2. `listen` sounds great to me.
3. Let's add a demo in the StarWars example. This way we can quickly test subscriptions it in new releases. I was thinking of a live clock in the form of subscriptions. I can work on the server here if necessary.
4. Do you feel like we should add unit tests for anything?
5. Let's add documentation with a small usage example - nothing fancy.

What do you think? What should I help you with?

**Reply by Amzd** [reply]
The additional send is because you cannot call a subscription with a normal http request (at least not in our environment lmk if this is not correct, I've never worked with graphql before). If you do not write the Subscription send call then you have to write an extra set of send methods and specify Query in one and Mutation in the other because swift generics are only inclusive not exclusive. 

~~I forgot that we can probably type check TypeLock to be Operations.Subscription can't we?~~


I'll see if I have some time tomorrow and otherwise I'll get something done over the weekend.

**Reply by Amzd** [reply]
I think we should have an HTTPOperation and WebSocketOperation protocol instead which would be a nicer way to implement the definition which call you can make with that operation; `send` or `listen`.

Is there any particular reason why there is no code shared between codegen and the project side target?

**Reply by maticzav** [reply]
Let’s do that! What do you mean by code shared?

**Comment by shaps80** [comment]
This is now fairly outdated and I'm just cleaning up. Sounds like it was mostly completed anyway, closing 👍
